lib/logstash/plugin_mixins/normalize_config_support.rb

```diff
@@ -0,0 +1,128 @@
+      #
+      # @yieldreturn a configurator block that can be used to configure the normalization
+      # @param canonical_config the config name to produce a normalized value for
+      def normalize_config(canonical_config, &configurator)
+        ConfigNormalizer.new(self, canonical_config, &configurator).value
+      end
+
+      class ConfigNormalizer
+        def initialize(plugin, canonical_config, &configurator)
+          require_value!('canonical_config', canonical_config)
+          require_value!('configurator', configurator)
+
+          @plugin = plugin
+          @canonical_config = canonical_config.to_s
+          @deprecated_mappings = nil
+          ensure_config_exists!(@canonical_config)
+
+          configurator.call(self)
+        end
+
+        ##
+        # Map one or more deprecated configs to the current canonical config.
+        #
+        # The `value_transformer` block is used when one or more deprecated configs are explicitly supplied,
+        # to transform their effective values into a single suitable value for use as-if it had been
+        # provided by the canonical config.
+        #
+        # @raise [RuntimeError] if no deprecation mappings are set for the current canonical config
+        # @raise [ArgumentError] if any deprecated params provided is not marked deprecated
+        # @raise [ArgumentError] if an invalid value_transformer in provided
+        # @param deprecated_params [String...]: the deprecated param names
+        # @return [void]
+        def with_deprecated_mapping(*deprecated_params, &value_transformer)
+          fail 'Deprecated mappings already configured for this config normalizer' unless @deprecated_mappings.nil?
+          require_value!('deprecated_params', deprecated_params)
+          ensure_deprecated!(deprecated_params)
+          check_value_transformer!(deprecated_params, value_transformer)
+
+          @deprecated_mappings = [deprecated_params.map(&:to_s), value_transformer]
+        end
+
+        ##
+        # Wholly-alias a deprecated config to the current canonical config.
+        # Both canonical and deprecated alias must accept the same set of values.
+        #
+        # @see #with_deprecated_mapping
+        #
+        def with_deprecated_alias(deprecated_alias)
+          with_deprecated_mapping(deprecated_alias) { |v| v }
+        end
+
+        ##
+        # Unambiguously extracts the effective configuration value from the canonical config and deprecated param mappings.
+        # @raise `LogStash::ConfigurationError` if both canonical and deprecated params are supplied
+        # @raise [RuntimeError] if no deprecation mappings are set for the current canonical config
+        #
+        # @return [Object]: the value of the canonical config param or an equivalent derived from provided deprecated params
+        def value
+          fail 'No deprecated mappings configured for this config normalizer' if @deprecated_mappings.nil?
```

Comment:
🤔 This failure is a detail of the _current_ feature-set, but is a good value-add.

If/when we add features to this config normalizer, we will need to fail in _fewer_ scenarios (e.g., when a plugin maintainer relies on a new feature), but all of the invocations that currently pass will need to continue passing).

That said, we should probably fail with an `ArgumentError` to indicate that the support adapter is being used incorrectly

```suggestion
          fail(ArgumentError, 'No deprecated mappings configured for this config normalizer') if @deprecated_mappings.nil?
```